godotengine/godot
# Replace hand-written pre-commit hooks with `pre-commit` Python tool

### How to test

```
pip install pre-commit
rm .git/hooks/*
pre-commit install
```

Do some changes that would fail tests, and try to commit, or run `pre-commit run` manually.

---

[`pre-commit`](https://pre-commit.com/) can be installed with pip, and configured in the Godot repo with `pre-commit install`. It can then easily be run both locally with `pre-commit run`, and on CI, in a cross-platform way.

This makes it much easier for contributors to set up pre-commit hooks, without having to manually copy files to their git folder.

- Supersedes #87003
- Supersedes #46235

I've included fixes for the issues I mentioned in https://github.com/godotengine/godot/pull/87003#issuecomment-1944326039:
- Merged the two `black` ids back together and made it apply on all `.py`, `SConstruct` and `SCsub` files, instead of relying on `types_or: [python]` which doesn't know SCons files.
- Replaced the custom clang-format wrapper with the official `pre-commit` hook at https://github.com/pre-commit/mirrors-clang-format, which takes care of installing the clang-format version we want via Python. This solves the need of ensuring everyone uses the same clang-format version installed locally.
- Removed all `misc/hooks/` files which are no longer needed.
- Removed `misc/scripts/{black,clang}_format.sh`, replaced by running `pre-commit run` locally.

---

@Calinou IMO some of the changes from #46235 may still be worth bringing over, maybe as a follow up PR. You seemed to have found pre-existing hooks that can help us get rid of `misc/scripts/file_format.sh`, and the global exclude pattern may not be a bad idea.

We can likely look into moving more of the `misc/scripts/` / CI checks to `pre-commit`, e.g. dotnet format (CC @godotengine/dotnet) or codespell.
